File: lib/ts/src/internal/binary/PackMap.ts
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

import { BinaryPackNode } from './BinaryPackNode';
import { CannotPack } from './CannotPack';
import { pack } from './Pack';
import { addExtension } from 'msgpackr';

const UNDEFINED_BYTE = 18;
export class MsgpackUndefined {
    toString() {
        return 'UNDEFINED';
    }
}
const MSGPACK_UNDEFINED_EXT = {
    Class: MsgpackUndefined,
    type: UNDEFINED_BYTE,
    pack(instance: MsgpackUndefined) {
        return Buffer.from([]);
    },
    unpack(buffer: Buffer) {
        return new MsgpackUndefined();
    },
};
addExtension(MSGPACK_UNDEFINED_EXT);

export function packMap(m: Map<any, any>, header: any[], keyIndexMap: Map<number, BinaryPackNode>): any[] {
    const row: any[] = [];
    for (const [key, value] of m.entries()) {
        if (typeof key === 'string') {
            throw new CannotPack();
        }

        const keyIndex = keyIndexMap.get(key);

        let finalKeyIndex: BinaryPackNode;
        if (keyIndex === undefined) {
            finalKeyIndex = new BinaryPackNode(header.length - 1, new Map());

            if (value instanceof Map) {
                header.push([key]);
            } else {
                header.push(key);
            }

            keyIndexMap.set(key, finalKeyIndex);
        } else {
            finalKeyIndex = keyIndex;
        }

        const keyIndexValue: number = finalKeyIndex.value;
        const keyIndexNested: Map<number, BinaryPackNode> = finalKeyIndex.nested;

        let packedValue: any;
        if (value instanceof Map && value !== null) {
            const nestedHeader: any[] = header[keyIndexValue + 1];
            if (!Array.isArray(nestedHeader)) {
                // No nesting available, so the data structure is inconsistent
                throw new CannotPack();
            }
            packedValue = packMap(value, nestedHeader, keyIndexNested);
        } else {
            if (Array.isArray(header[keyIndexValue + 1])) {
                throw new CannotPack();
            }

            packedValue = pack(value);
        }

        while (row.length < keyIndexValue) {
            row.push(new MsgpackUndefined());
        }

        if (row.length === keyIndexValue) {
            row.push(packedValue);
        } else {
            row[keyIndexValue] = packedValue;
        }
    }
    return row;
}

File: lib/go/internal/binary/PackMap.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package binary

import "sort"

// PackMap attempts to pack a dictionary with integer keys into a compact row representation.
func PackMap(m map[int]any, header *[]any, keyIndexMap map[int]*BinaryPackNode) ([]any, error) {
	row := make([]any, 0, len(m))

	keys := make([]int, 0, len(m))
	for key := range m {
		keys = append(keys, key)
	}
	sort.Ints(keys)

	for _, key := range keys {
		rawValue := m[key]
		node, exists := keyIndexMap[key]
		if !exists {
			node = NewBinaryPackNode(len(*header)-1, make(map[int]*BinaryPackNode))

			if _, ok := toIntKeyMap(rawValue); ok {
				*header = append(*header, []any{key})
			} else {
				*header = append(*header, key)
			}

			keyIndexMap[key] = node
		}

		index := node.Value

		if mapValue, ok := toIntKeyMap(rawValue); ok {
			nestedHeaderAny := (*header)[index+1]
			nestedHeader, ok := nestedHeaderAny.([]any)
			if !ok {
				return nil, CannotPack{}
			}

			nestedMap := node.Nested
			if nestedMap == nil {
				nestedMap = make(map[int]*BinaryPackNode)
				node.Nested = nestedMap
			}

			nestedHeaderCopy := nestedHeader
			nestedRow, err := PackMap(mapValue, &nestedHeaderCopy, nestedMap)
			if err != nil {
				return nil, err
			}
			(*header)[index+1] = nestedHeaderCopy

			row = ensureRowCapacity(row, index)
			if len(row) == index {
				row = append(row, nestedRow)
			} else {
				row[index] = nestedRow
			}
		} else {
			if _, ok := (*header)[index+1].([]any); ok {
				return nil, CannotPack{}
			}

			packedValue, err := Pack(rawValue)
			if err != nil {
				return nil, err
			}

			row = ensureRowCapacity(row, index)
			if len(row) == index {
				row = append(row, packedValue)
			} else {
				row[index] = packedValue
			}
		}
	}

	return row, nil
}

func ensureRowCapacity(row []any, index int) []any {
	for len(row) < index {
		row = append(row, undefinedMarker)
	}
	return row
}
